GPflow glossary document (#923)
GPflow/GPflow: GLOSSARY.md

> <dl>

Reviewer: Would be good to have a header & brief explanation. E.g.:

```
Glossary
========

GPflow does not always follow standard Python naming conventions,
and instead tries to follow the notation in the relevant GP papers.
The following is the convention we aim to follow in the code:
```

  <dt>GPR</dt>

>   <dd>gaussian process regression</dd>

Reviewer: ```suggestion
  <dd>Gaussian process regression</dd>
```

  <dt>SVGP</dt>
  <dd>stochastic variational inference for Gaussian process models</dd>

>   <dt>[] shape constructions</dt>

Reviewer: ```suggestion
  <dt>`[...]` shape annotations</dt>
```

Author: @st-- , it is slightly misleading because later we say that the ellipsis sign `...` denotes leading dimensions.

Author: @st-- , but I agree that empty brackets don't tell much as well.

>   <dd>the way of describing tensor shape in docstrings and comments. Example: [..., N, D, D], this is tensor with unknown number of leading dimensions and last two dimensions are equal</dd>

Reviewer: ```suggestion
  <dd>The way of describing tensor shapes in docstrings and comments. Example: [..., N, D, D], this is a tensor with an arbitrary number of leading dimensions, and the last two dimensions are equal</dd>
```

  <dt>X</dt>

>   <dd>and its variations like Xnew in GPflow code mean input data (numpy or tensorflow)</dd>

Reviewer: ```suggestion
  <dd>(and variations like Xnew) refers to input points; always of rank 2, e.g. shape [N, D], even when D=1</dd>
```
(I think a lot of the entries can describe both numpy arrays or tensorflow tensors so I think it's confusing if this is the only place where it's mentioned explicitly)

  <dt>Y</dt>

>   <dd>throughout is used to denote observed data, potentially with multiple output dimensions</dd>

Reviewer: ```suggestion
  <dd>refers to observed output values, potentially with multiple output dimensions; always of rank 2, e.g. shape [N,
 P], even when P=1</dd>
```
Would it be useful to point towards the notebooks here (Coregionalization still has "P=1", the outputs are distinguished by one of the columns in the input points)?

  <dt>N</dt>
  <dd>stands for data or minibatch size in docstrings and shape constructions</dd>

  <dt>P</dt>
  <dd>stands for the number of output dimensions in docstrings and shape constructions</dd>

  <dt>D</dt>
  <dd>stands for the number of input dimensions in docstrings and shape constructions</dd>

  <dt>Z</dt>

>   <dd>inducing points tensor</dd>

Reviewer: ```suggestion
  <dd>inducing points</dd>
```

  <dt>...</dt>

>   <dd>leading dimensions in docstrings and in the code it is ellipsis.</dd>

Reviewer: ```suggestion
  <dd>arbitrary leading dimensions</dd>
```
(this is true both for docstrings and code)

Reviewer: maybe move this right underneath the entry for shapes ?

## Glossary

GPflow does not always follow standard Python naming conventions,
and instead tries to apply the notation in the relevant GP papers.\
The following is the convention we aim to use in the code.

---

<dl>
  <dt>GPR</dt>
  <dd>Gaussian process regression</dd>

  <dt>SVGP</dt>
  <dd>stochastic variational inference for Gaussian process models</dd>

  <dt>Shape constructions [..., A, B]</dt>
  <dd>the way of describing tensor shapes in docstrings and comments. Example: <i>[..., N, D, D]</i>, this is a tensor with an arbitrary number of leading dimensions indicated using the ellipsis sign, and the last two dimensions are equal</dd>

  <dt>X</dt>
  <dd>(and variations like Xnew) refers to input points; always of rank 2, e.g. shape <i>[N, D]</i>, even when <i>D=1</i></dd>

  <dt>Y</dt>
  <dd>(and variations like Ynew) refers to observed output values, potentially with multiple output dimensions; always of rank 2, e.g. shape <i>[N, P]</i>, even when <i>P=1</i></dd>

  <dt>Z</dt>
  <dd>refers to inducing points</dd>

> 

Reviewer: ```suggestion

  <dt>M</dt>
  <dd>stands for the number of inducing features (e.g. length of Z)</dd>

```

## Glossary

GPflow does not always follow standard Python naming conventions,
and instead tries to apply the notation in the relevant GP papers.\
The following is the convention we aim to use in the code.

---

<dl>
  <dt>GPR</dt>
  <dd>Gaussian process regression</dd>

  <dt>SVGP</dt>
  <dd>stochastic variational inference for Gaussian process models</dd>

  <dt>Shape constructions [..., A, B]</dt>
  <dd>the way of describing tensor shapes in docstrings and comments. Example: <i>[..., N, D, D]</i>, this is a tensor with an arbitrary number of leading dimensions indicated using the ellipsis sign, and the last two dimensions are equal</dd>

  <dt>X</dt>
  <dd>(and variations like Xnew) refers to input points; always of rank 2, e.g. shape <i>[N, D]</i>, even when <i>D=1</i></dd>

  <dt>Y</dt>
  <dd>(and variations like Ynew) refers to observed output values, potentially with multiple output dimensions; always of rank 2, e.g. shape <i>[N, P]</i>, even when <i>P=1</i></dd>

  <dt>Z</dt>
  <dd>refers to inducing points</dd>

  <dt>M</dt>
  <dd>stands for the number of inducing features (e.g. length of Z)</dd>

  <dt>N</dt>

>   <dd>stands for data or minibatch size in docstrings and shape constructions</dd>

Author: ```suggestion
  <dd>stands for the number of data or minibatch size in docstrings and shape constructions</dd>
```
